JOSESwift/Sources/Decrypter.swift

```diff
@@ -23,140 +23,70 @@
-    ///
-    /// - Parameters:
-    ///   - context: The `SymmetricDecryptionContext` containing the ciphertext, the initialization vector, additional authenticated data and the authentication tag.
-    ///   - symmetricKey: The key which contains the HMAC and encryption key.
-    /// - Returns: The plain text (decrypted cipher text).
-    /// - Throws: `EncryptionError` if any error occurs during decryption.
-    func decrypt(_ context: SymmetricDecryptionContext, with symmetricKey: Data) throws -> Data
-}
-
-public struct DecryptionContext {
-    let header: JWEHeader
-    let encryptedKey: Data
-    let initializationVector: Data
-    let ciphertext: Data
-    let authenticationTag: Data
-}
-
-public struct SymmetricDecryptionContext {
-    let ciphertext: Data
-    let initializationVector: Data
-    let additionalAuthenticatedData: Data
-    let authenticationTag: Data
-}
-
 public struct Decrypter {
-    let asymmetric: AsymmetricDecrypter
-    let symmetric: SymmetricDecrypter
-
-    /// Constructs a decrypter used to decrypt a JWE.
-    ///
-    /// - Parameters:
-    ///   - keyDecryptionAlgorithm: The algorithm used to decrypt the shared content encryption key.
-    ///   - key: The key used to perform the decryption. If the `keyDecryptionAlgorithm` is `.direct`, the
-    ///          `decryptionKey` is the shared symmetric content encryption key. Otherwise the `decryptionKey` is the
-    ///           private key of the receiver. See [RFC-7516](https://tools.ietf.org/html/rfc7516#section-5.2) for
-    ///           details.
-    ///   - contentDecryptionAlgorithm: The algorithm used to decrypt the JWE's payload.
-    /// - Returns: A fully initialized `Decrypter` or `nil` if provided key is of the wrong type.
-    public init?<KeyType>(keyDecryptionAlgorithm: KeyManagementAlgorithm, decryptionKey key: KeyType, contentDecryptionAlgorithm: ContentEncryptionAlgorithm) {
-        // TODO: This switch won't scale. We need to refactor it. (#141)
-        switch (keyDecryptionAlgorithm, contentDecryptionAlgorithm) {
-        case (.RSA1_5, .A256CBCHS512), (.RSAOAEP, .A256CBCHS512), (.RSAOAEP256, .A256CBCHS512), (.RSA1_5, .A128CBCHS256), (.RSAOAEP, .A128CBCHS256), (.RSAOAEP256, .A128CBCHS256):
-            guard type(of: key) is RSADecrypter.KeyType.Type else {
-                return nil
-            }
-            // swiftlint:disable:next force_cast
-            self.asymmetric = RSADecrypter(algorithm: keyDecryptionAlgorithm, privateKey: (key as! RSADecrypter.KeyType))
-            self.symmetric = AESDecrypter(algorithm: contentDecryptionAlgorithm)
-        case (.direct, .A256CBCHS512), (.direct, .A128CBCHS256):
-            guard type(of: key) is AESDecrypter.KeyType.Type else {
-                return nil
-            }
+    let keyManagementMode: DecryptionKeyManagementMode
+    let keyManagementAlgorithm: KeyManagementAlgorithm
+    let contentEncryptionAlgorithm: ContentEncryptionAlgorithm
+
+    public init?<KeyType>(
```

Comment:
The original idea was to keep any crypto implementation out of that part. But you have a good point there. I think it might actually be more understandable to users if they knew that the type must either be ` SecKey` or `Data`.  I have to update the documentation the public initializers anyway. I'll constraint the generic type or add at least a clear point to the documentation then.